cmd/kubeadm/app/phases/upgrade/staticpods_test.go

```diff
@@ -225,24 +226,44 @@ func (spm *fakeStaticPodPathManager) CleanupDirs() error {
 	return os.RemoveAll(spm.BackupEtcdDir())
 }
 
-type fakeTLSEtcdClient struct{ TLS bool }
+type fakeTLSEtcdClient struct {
+	TLS              bool
+	version          string
+	versionErr       bool
+	clusterStatusErr bool
+	availableErr     bool
+}
 
 func (c fakeTLSEtcdClient) ClusterAvailable() (bool, error) { return true, nil }
 
 func (c fakeTLSEtcdClient) WaitForClusterAvailable(retries int, retryInterval time.Duration) (bool, error) {
+	if c.availableErr {
+		return false, errors.New("WaitForClusterAvailable failed")
+	}
 	return true, nil
 }
 
 func (c fakeTLSEtcdClient) GetClusterStatus() (map[string]*clientv3.StatusResponse, error) {
+	if c.clusterStatusErr {
+		return nil, errors.New("GetClusterStatus failed")
+	}
 	return map[string]*clientv3.StatusResponse{
-		"https://1.2.3.4:2379": {
+		"https://192.168.2.2:2379": {
```

Comment:
There used "192.168.2.2" in const testConfiguration, so this place should use "192.168.2.2", otherwise it will return error when currentEtcdVersions[etcdutil.GetClientURL(&cfg.LocalAPIEndpoint)]. The original test case used “1.2.3.4” and success because the following code omitted this error.

currentEtcdVersionStr, ok := currentEtcdVersions[etcdutil.GetClientURL(&cfg.LocalAPIEndpoint)]
	if !ok {
		return true, errors.Wrap(err, "failed to retrieve the current etcd version from cfg.LocalAPIEndpoint")
	}

Here err is nil, and errors.Wrap(err, "failed to retrieve the current etcd version from cfg.LocalAPIEndpoint") return nil, I think it should be as following:

currentEtcdVersionStr, ok := currentEtcdVersions[etcdutil.GetClientURL(&cfg.LocalAPIEndpoint)]
	if !ok {
		return true, errors.Wrap(errors.Errorf("unknown LocalAPIEndpoint: %v", cfg.LocalAPIEndpoint), "failed to retrieve the current etcd version from cfg.LocalAPIEndpoint")
	}

Reply:
We can have that IP as a field in `fakeTLSEtcdClient` and fill it in from the parsed config or in a const. I am definitely not a fan of copy-pasting it everywhere in this file.